ops/testing.py

```diff
@@ -389,57 +389,104 @@ def cleanup(self) -> None:
         """
         self._backend._cleanup()
 
-    def _create_meta(self, charm_metadata: Optional[YAMLStringOrFile],
-                     action_metadata: Optional[YAMLStringOrFile]) -> CharmMeta:
+    def _create_meta(self, charm_metadata_yaml: Optional[YAMLStringOrFile],
+                     action_metadata_yaml: Optional[YAMLStringOrFile]) -> CharmMeta:
         """Create a CharmMeta object.
 
         Handle the cases where a user doesn't supply explicit metadata snippets.
+        This will try to load metadata from ``<charm_dir>/charmcraft.yaml`` first, then
+        ``<charm_dir>/metadata.yaml`` if charmcraft.yaml does not include metadata,
+        and ``<charm_dir>/actions.yaml`` if charmcraft.yaml does not include actions.
         """
         filename = inspect.getfile(self._charm_cls)
         charm_dir = pathlib.Path(filename).parents[1]
 
+        charm_metadata: Optional[Dict[str, Any]] = None
+        charmcraft_metadata: Optional[dict[str, Any]] = None
+        # Check charmcraft.yaml and load it if it exists
+        charmcraft_meta = charm_dir / "charmcraft.yaml"
+        if charmcraft_meta.is_file():
+            self._charm_dir = charm_dir
+            charmcraft_metadata = yaml.safe_load(charmcraft_meta.read_text())
+
+        # Load metadata from parameters if provided
+        if charm_metadata_yaml is not None:
+            if isinstance(charm_metadata_yaml, str):
+                charm_metadata_yaml = dedent(charm_metadata_yaml)
+            charm_metadata = yaml.safe_load(charm_metadata_yaml)
+        else:
+            # Check charmcraft.yaml for metadata if no metadata is provided
+            if charmcraft_metadata is not None:
+                if any(
+                    meta_key in charmcraft_metadata
+                    for meta_key in ("name", "summary", "description")
+                ):
```

Comment:
Style nit, but arguably a bit easier to read on two lines than four (and this is all about "meta", so no need for the `meta_` prefix in the loop variable name):


```suggestion
                meta_keys = ["name", "summary", "description"]
                if any(key in charmcraft_metadata for key in meta_keys):
```